fendermint/testing/contract-test/src/lib.rs

```diff
@@ -56,3 +69,182 @@ pub async fn init_exec_state(
+            base_fee: out.base_fee,
+            circ_supply: out.circ_supply,
+            chain_id: out.chain_id.into(),
+            power_scale: out.power_scale,
+            app_version: 0,
+        };
+
+        Ok(())
+    }
+
+    /// Take the execution state, update it, put it back, return the output.
+    async fn modify_exec_state<T, F, R>(&self, f: F) -> anyhow::Result<T>
+    where
+        F: FnOnce(FvmExecState<MemoryBlockstore>) -> R,
+        R: Future<Output = Result<(FvmExecState<MemoryBlockstore>, T)>>,
+    {
+        let mut guard = self.exec_state.lock().await;
+        let state = guard.take().expect("exec state empty");
+
+        let (state, ret) = f(state).await?;
+
+        *guard = Some(state);
+
+        Ok(ret)
+    }
+
+    /// Put the execution state during block execution. Has to be empty.
+    async fn put_exec_state(&self, state: FvmExecState<MemoryBlockstore>) {
+        let mut guard = self.exec_state.lock().await;
+        assert!(guard.is_none(), "exec state not empty");
+        *guard = Some(state);
+    }
+
+    /// Take the execution state during block execution. Has to be non-empty.
+    async fn take_exec_state(&self) -> FvmExecState<MemoryBlockstore> {
+        let mut guard = self.exec_state.lock().await;
+        guard.take().expect("exec state empty")
+    }
+
+    pub async fn begin_block(&self, block_height: ChainEpoch) -> Result<()> {
+        // TODO: generate block hash based on input
+        let block_hash: [u8; 32] = [0; 32];
+
+        let db = self.state_store.as_ref().clone();
+        let mut state_params = self.state_params.clone();
+        state_params.timestamp = Timestamp(block_height as u64);
+
+        let state = FvmExecState::new(db, self.multi_engine.as_ref(), block_height, state_params)
+            .context("error creating new state")?
+            .with_block_hash(block_hash);
+
+        self.put_exec_state(state).await;
+
+        let _res = self
+            .modify_exec_state(|s| self.interpreter.begin(s))
+            .await
+            .unwrap();
+
+        Ok(())
+    }
+
+    pub async fn end_block(&self, _block_height: ChainEpoch) -> Result<()> {
+        let _ret = self
+            .modify_exec_state(|s| self.interpreter.end(s))
+            .await
+            .context("end failed")?;
+
+        Ok(())
+    }
+
+    pub async fn commit(&mut self) -> Result<()> {
+        let exec_state = self.take_exec_state().await;
+
+        let (
+            state_root,
+            FvmUpdatableParams {
+                power_scale,
+                circ_supply,
+                app_version,
+            },
+            _,
+        ) = exec_state.commit().context("failed to commit FVM")?;
+
+        self.state_params.state_root = state_root;
+        self.state_params.power_scale = power_scale;
+        self.state_params.circ_supply = circ_supply;
+        self.state_params.app_version = app_version;
+
+        println!("self.state_params: {:?}", self.state_params);
```

Comment:
```suggestion
        eprintln!("self.state_params: {:?}", self.state_params);
```
I tend to use `eprintln!` for debug logs in tests, so `println!` can be reserved for stuff you really want as output, in general.